common/channel/src/message_queues.rs

```diff
@@ -139,18 +154,51 @@ impl<K: Eq + Hash + Clone, T> PerKeyQueue<K, T> {
                 return None;
             }
         };
+
         let (message, is_q_empty) = self.pop_from_key_queue(&key);
         if !is_q_empty {
             self.round_robin_queue.push_back(key);
         }
+
         if message.is_some() {
             if let Some(c) = self.counters.as_ref() {
                 c.with_label_values(&["dequeued"]).inc();
             }
+
+            // Remove empty per-key-queues every `POPS_PER_GC` successful dequeue
+            // operations.
+            //
+            // libra-channel never removes keys from its PerKeyQueue (without
+            // this logic). This works fine for the validator network, where we
+            // have a bounded set of peers that almost never changes; however,
+            // this does not work for servicing public clients, where we can have
+            // large and frequent connection churn.
+            //
+            // Periodically removing these empty queues prevents us from causing
+            // an effective memory leak when we have lots of transient peers in
+            // e.g. the public-facing vfn use-case.
+            //
+            // This GC strategy could probably be more sophisticated, though it
+            // seems to work well in some basic stress tests / micro benches.
+            //
+            // See: common/channel/src/bin/many_keys_stress_test.rs
+            //
+            // For more context, see: https://github.com/libra/libra/issues/5543
+            self.num_popped_since_gc += 1;
+            if self.num_popped_since_gc == POPS_PER_GC {
+                self.num_popped_since_gc = 0;
+                self.remove_empty_queues();
+            }
         }
+
         message
     }
 
+    /// Garbage collect any empty per-key-queues.
+    fn remove_empty_queues(&mut self) {
+        self.per_key_queue.retain(|_key, queue| !queue.is_empty());
+    }
```

Comment:
I'm not fully up to date with how this works, but I saw earlier you add the key to the round robin queue, should you remove the key from the round robin queue?

Reply:
iirc, key ∈ round_robin_queue <==> !per_key_queue[key].is_empty()

so all empty queues will not have their keys in the round_robin_queue